openqaoa/optimizers/optimization_methods.py

```diff
@@ -431,3 +436,161 @@ def stochastic_grad_descent(fun, x0, jac, args=(), stepsize=0.001, mass=0.9, sta
+    testy = besty
+    while improved and not stop and niter < maxiter:
+        t0 = time.time()
+        
+        print('a',n_shots,testy)
+
+        # compute gradient and variance
+        gradient, variance = jac_w_variance(testx, n_shots=n_shots)
+
+        # add the number of shots to the total
+        n_shots_total += 2*n_shots*testx.size
+
+        # compute gradient descent step
+        testx = testx - stepsize*gradient
+        testy = np.real(fun(testx, *args))
+
+        # compute n_shots for next step
+        chi = mu*chi + (1-mu)*gradient
+        xi = mu*xi + (1-mu)*np.sum(variance)        
+        n_shots = int(np.ceil(2*lipschitz*stepsize*xi/((2-lipschitz*stepsize)*(np.linalg.norm(chi)**2+b*mu**niter))))
+
+        # clip the number of shots
+        n_shots = max(n_shots, n_shots_min)
+        n_shots = min(n_shots, n_shots_max) if n_shots_max else n_shots
+     
+
+        if np.abs(besty-testy) < tol:
+            improved = False
+
+        else:
+            besty = testy
+            bestx = np.copy(testx)
+            improved = True
+
+        if callback is not None:
+            callback(bestx)
+        if maxfev is not None and funcalls >= maxfev:
+            stop = True
+            break
+
+        # if there is a maximum number of shots, check if the remaining shots are enough for next step, otherwise stop
+        if n_shots_total_max != None:
+            if n_shots_total_max - n_shots_total < 2*n_shots*testx.size:
+                stop = True
+                break
+
+        niter += 1
+        print('iter',niter,'time',time.time()-t0)
+
+    return OptimizeResult(fun=besty, x=bestx, nit=niter,
+                          nfev=funcalls, success=(niter > 1))
+
+
+def iCANS(fun, x0, args=(), maxfev=None, stepsize=0.00001, n_shots_min=10, n_shots_max=10000, n_shots_total_max=None, 
+            mu=0.99, b=1e-06, coeffs=None, maxiter=100, tol=10**(-6), jac_w_variance=None,callback=None, **options): 
+
+    # check that the stepsize is small enough
+    lipschitz = np.sum(np.abs(coeffs)) 
+    if not stepsize < 2/lipschitz:
+        raise ValueError("Stepsizec is bigger than 2/Lipschitz: it should be smaller than {0:.3g}".format(2/lipschitz))
+
+    # define the initial values
+    chi_ = np.zeros(len(x0))
+    xi_ = np.zeros(len(x0))
+    n_shots = [n_shots_min for _ in range(len(x0))]
+
+    bestx = x0
+    besty = np.real(fun(bestx, *args))
+    funcalls = 1  # tracks no. of function evals.
+    niter = 0
+    n_shots_total = 0 # track no. of shots
+    improved = True
+    stop = False
+
+    testx = np.copy(bestx)
+    testy = besty
+    while improved and not stop and niter < maxiter:
+
+        t0 = time.time()
+        
+        print('a',n_shots,testy)
+
+        # compute gradient and variance
+        gradient, variance = jac_w_variance(testx, n_shots=list(n_shots))
+        
+        # add the number of shots to the total
+        n_shots_total += 2 * np.sum(n_shots, dtype=int)
```

Comment:
Do we want to be able to put a budget of number of shots? So if `n_shots_total > n_shots_budget` then we stop the optmization loop